<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Data Wrangling and Processing for Transcriptomics: Assessing Read Quality</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/02-quality-control.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Data Wrangling and Processing for Transcriptomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Data Wrangling and Processing for Transcriptomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Data Wrangling and Processing for Transcriptomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 25%" class="percentage">
    25%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 25%" aria-valuenow="25" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/02-quality-control.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-background.html">1. Introduction and Experimental Planning Considerations for RNA-Seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        2. Assessing Read Quality
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#bioinformatic-workflows">Bioinformatic workflows</a></li>
<li><a href="#starting-with-data">Starting with data</a></li>
<li><a href="#quality-control">Quality control</a></li>
<li><a href="#other-notes-optional">Other notes – optional</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-trimming.html">3. Trimming and Filtering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-variant_calling.html">4. Variant Calling Workflow</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-automation.html">5. Automating a Variant Calling Workflow</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li><li><a href="reference.html">Glossary</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="01-background.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="03-trimming.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="01-background.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction and
        </a>
        <a class="chapter-link float-end" href="03-trimming.html" rel="next">
          Next: Trimming and...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Assessing Read Quality</h1>
        <p>Last updated on 2024-08-19 |

        <a href="https://github.com/btmoyers/wrangling-transcriptomics/edit/main/episodes/02-quality-control.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can I describe the quality of my data?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain how a FASTQ file encodes per-base quality scores.</li>
<li>Interpret a FastQC plot summarizing per-base quality across all
reads.</li>
<li>Use <code>for</code> loops to automate operations on multiple
files.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="bioinformatic-workflows">Bioinformatic workflows<a class="anchor" aria-label="anchor" href="#bioinformatic-workflows"></a>
<a class="anchor" aria-label="anchor" href="#bioinformatic-workflows"></a></h2>
<hr class="half-width"><p>When working with high-throughput sequencing data, the raw reads you
get off of the sequencer will need to pass through a number of different
tools in order to generate your final desired output. The execution of
this set of tools in a specified order is commonly referred to as a
<em>workflow</em> or a <em>pipeline</em>.</p>
<p>An example of the workflow we will be using for our variant calling
analysis is provided below with a brief description of each step.</p>
<figure><img src="fig/variant_calling_workflow.png" alt="workflow" class="figure mx-auto d-block"></figure><ol style="list-style-type: decimal"><li>Quality control - Assessing quality using FastQC</li>
<li>Quality control - Trimming and/or filtering reads (if
necessary)</li>
<li>Align reads to reference genome</li>
<li>Perform post-alignment clean-up</li>
<li>Variant calling</li>
</ol><p>These workflows in bioinformatics adopt a plug-and-play approach in
that the output of one tool can be easily used as input to another tool
without any extensive configuration. Having standards for data formats
is what makes this feasible. Standards ensure that data is stored in a
way that is generally accepted and agreed upon within the community. The
tools that are used to analyze data at different stages of the workflow
are therefore built under the assumption that the data will be provided
in a specific format.</p>
</section><section><h2 class="section-heading" id="starting-with-data">Starting with data<a class="anchor" aria-label="anchor" href="#starting-with-data"></a>
<a class="anchor" aria-label="anchor" href="#starting-with-data"></a></h2>
<hr class="half-width"><p>Our data are three samples from an experiment on the conversion of
fibroblasts to myofibroblasts, <a href="https://www.doi.org/10.1038/s41598-018-21506-7" class="external-link">Patalano et
al. 2018</a>. There are two known signalling proteins involved in the
conversion of prostate fibroblasts to myofibroblasts: TBF-beta and
CXCL12. These proteins both affect which genes get transcribed in cells,
but they act on different pathways and do not affect all of the same
genes as each other.</p>
<p>In this experiment, fibroblasts were treated with either TGF-beta,
CXCL12, or a control (termed ‘vehicle’). The first letter of each sample
name corresponds to the treatment (eg. the samples treated with TGF-beta
begin with T and the vehicle controls begin with a V). For each sample,
we have one file with all the forward reads and one file with all the
reverse reads.</p>
</section><section><h2 class="section-heading" id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
<a class="anchor" aria-label="anchor" href="#quality-control"></a></h2>
<hr class="half-width"><p>We will now assess the quality of the sequence reads contained in our
fastq files.</p>
<figure><img src="fig/var_calling_workflow_qc.png" alt="workflow_qc" class="figure mx-auto d-block"></figure><div class="section level3">
<h3 id="details-on-the-fastq-format">Details on the FASTQ format<a class="anchor" aria-label="anchor" href="#details-on-the-fastq-format"></a></h3>
<p>Although it looks complicated (and it is), we can understand the <a href="https://en.wikipedia.org/wiki/FASTQ_format" class="external-link">fastq</a> format with
a little decoding. Some rules about the format include…</p>
<table class="table"><colgroup><col width="3%"><col width="96%"></colgroup><thead><tr class="header"><th>Line</th>
<th>Description</th>
</tr></thead><tbody><tr class="odd"><td>1</td>
<td>Always begins with ‘@’ and then information about the read</td>
</tr><tr class="even"><td>2</td>
<td>The actual DNA sequence</td>
</tr><tr class="odd"><td>3</td>
<td>Always begins with a ‘+’ and sometimes the same info in line 1</td>
</tr><tr class="even"><td>4</td>
<td>Has a string of characters which represent the quality scores; must
have same number of characters as line 2</td>
</tr></tbody></table><p>We can view the first complete read in one of the files our dataset
by using <code>head</code> to look at the first four lines.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-n</span> 4 C1_S4_L001_R2_001_downsampled.fastq</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@D00345:37:HBATBADXX:1:1214:3724:1975 2:N:0:GCCAAT
NNNNNNNAATCNTTNCAANTCTCTTGCAAGGTNCCCTGGTTGNGAAAATAC
+
#######22:@#2:#22&gt;#2@3=?=&lt;?@??&lt;&lt;#0:;=??&gt;&gt;?#0-=??&gt;&gt;=</code></pre>
</div>
<p>Line 4 shows the quality for each nucleotide in the read. Quality is
interpreted as the probability of an incorrect base call (e.g. 1 in 10)
or, equivalently, the base call accuracy (e.g. 90%). To make it possible
to line up each individual nucleotide with its quality score, the
numerical score is converted into a code where each individual character
represents the numerical quality score for an individual nucleotide. For
example, in the line above, the quality score line is:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="co">#######22:@#2:#22&gt;#2@3=?=&lt;?@??&lt;&lt;#0:;=??&gt;&gt;?#0-=??&gt;&gt;=</span></span></code></pre>
</div>
<p>The numerical value assigned to each of these characters depends on
the sequencing platform that generated the reads. The sequencing machine
used to generate our data uses the standard Sanger quality PHRED score
encoding, using Illumina version 1.8 onwards. Each character is assigned
a quality score between 0 and 41 as shown in the chart below.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Quality encoding: !"#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHIJ
                   |         |         |         |         |
Quality score:    01........11........21........31........41</code></pre>
</div>
<p>Each quality score represents the probability that the corresponding
nucleotide call is incorrect. This quality score is logarithmically
based, so a quality score of 10 reflects a base call accuracy of 90%,
but a quality score of 20 reflects a base call accuracy of 99%. These
probability values are the results from the base calling algorithm and
depend on how much signal was captured for the base incorporation.</p>
<p>Looking back at our read:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@D00345:37:HBATBADXX:1:1214:3724:1975 2:N:0:GCCAAT
NNNNNNNAATCNTTNCAANTCTCTTGCAAGGTNCCCTGGTTGNGAAAATAC
+
#######22:@#2:#22&gt;#2@3=?=&lt;?@??&lt;&lt;#0:;=??&gt;&gt;?#0-=??&gt;&gt;=</code></pre>
</div>
<p>We can now see that there is a range of quality scores, but that the
start of the sequence is very poor (<code>#</code> = a quality score of
2).</p>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise"></a>
</h3>
<div class="callout-content">
<p>What is the last read in the
<code>V1_S1_L001_R1_001_downsampled.fastq</code> file? How confident are
you in this read?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">$</span> tail <span class="at">-n</span> 4 V1_S1_L001_R1_001_downsampled.fastq</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@D00345:37:HBATBADXX:1:1206:21030:101413 1:N:0:CGATGT
GTTACTCGACCGAAGTCTTCACTATGCATCACAACTCAAGATTANNNTANA
+
@@@FDDFFHGDHHIGFH@@GGHGIIIIIGGDCEGGFHHGFGGAH###--#-
</code></pre>
</div>
<p>This read has higher quality quality overall than the first read that
we looked at, but still has a range of quality scores, including
low-quality bases at the end. We will look at variations in
position-based quality in just a moment.</p>
</div>
</div>
</div>
</div>
<p>At this point, let’s validate that all the relevant tools are
installed. If you are using the Chimera then these <em>should</em> be
pre-installed, but you’ll need to load some modules.</p>
<p>Why do we need to load modules? Chimera is a shared resource used by
many different people working on many different types of projects. Not
everybody needs to be able to use the same programs as each other, and
if we had every program loaded at the same time it would potentially
slow the chimera down and lead to conflicts between programs. Let’s take
a look at how many different modules exist on the chimera:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">$</span> module avail</span></code></pre>
</div>
<p>That was a lot!! Remember you can use <code>clear</code> to calm your
terminal down. Let’s look for just the module we need, by typing its
name after <code>module avail</code>.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="ex">$</span> module avail fastqc</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="ex">-----------------------</span> /share/apps/modulefiles/modules <span class="at">------------------------</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>   <span class="ex">fastqc-0.11.9-gcc-10.2.0-osi6pqc</span> <span class="er">(</span><span class="ex">L</span><span class="kw">)</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>  <span class="ex">Where:</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>   <span class="ex">L:</span>  Module is loaded</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="ex">Use</span> <span class="st">"module spider"</span> to find all possible modules and extensions.</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="ex">Use</span> <span class="st">"module keyword key1 key2 ..."</span> to search for all possible modules matching</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="ex">any</span> of the <span class="st">"keys"</span>.</span></code></pre>
</div>
<p>If FastQC is not installed then you would expect to see an error
like:</p>
<pre><code>$ module avail fastqc
No module(s) or extension(s) found!
Use "module spider" to find all possible modules and extensions.
Use "module keyword key1 key2 ..." to search for all possible modules matching
any of the "keys".
</code></pre>
<p>If this happens check with chimera’s manager <a href="https://www.umb.edu/rc/staff/" class="external-link">Jeff Dusenberry</a> before trying
to install it.</p>
<p>Now let’s load the FastQC module:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">$</span> module load fastqc-0.11.9-gcc-10.2.0-osi6pqc</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="assessing-quality-using-fastqc">Assessing quality using FastQC<a class="anchor" aria-label="anchor" href="#assessing-quality-using-fastqc"></a></h3>
<p>In real life, you will not be assessing the quality of your reads by
visually inspecting your FASTQ files. Rather, you will be using a
software program to assess read quality and filter out poor quality
reads. We will first use a program called <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
to visualize the quality of our reads. Later in our workflow, we will
use another program to filter out poor quality reads.</p>
<p>FastQC has a number of features which can give you a quick impression
of any problems your data may have, so you can take these issues into
consideration before moving forward with your analyses. Rather than
looking at quality scores for each individual read, FastQC looks at
quality collectively across all reads within a sample. The image below
shows one FastQC-generated plot that indicates a very high quality
sample:</p>
<figure><img src="fig/good_quality1.8.png" alt="good_quality" class="figure mx-auto d-block"></figure><p>The x-axis displays the base position in the read, and the y-axis
shows quality scores. In this example, the sample contains reads that
are 40 bp long. This is much shorter than the reads we are working with
in our workflow. For each position, there is a box-and-whisker plot
showing the distribution of quality scores for all reads at that
position. The horizontal red line indicates the median quality score and
the yellow box shows the 1st to 3rd quartile range. This means that 50%
of reads have a quality score that falls within the range of the yellow
box at that position. The whiskers show the absolute range, which covers
the lowest (0th quartile) to highest (4th quartile) values.</p>
<p>For each position in this sample, the quality values do not drop much
lower than 32. This is a high quality score. The plot background is also
color-coded to identify good (green), acceptable (yellow), and bad (red)
quality scores.</p>
<p>Now let’s take a look at a quality plot on the other end of the
spectrum.</p>
<figure><img src="fig/bad_quality1.8.png" alt="bad_quality" class="figure mx-auto d-block"></figure><p>Here, we see positions within the read in which the boxes span a much
wider range. Also, quality scores drop quite low into the “bad” range,
particularly on the tail end of the reads. The FastQC tool produces
several other diagnostic plots to assess sample quality, in addition to
the one plotted above.</p>
</div>
<div class="section level3">
<h3 id="running-fastqc">Running FastQC<a class="anchor" aria-label="anchor" href="#running-fastqc"></a></h3>
<p>We will now assess the quality of the reads that we downloaded.
First, make sure you are still in the new <code>untrimmed_fastq</code>
directory inside of your project’s data directory. Note that you may
have named this directory something different.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">$</span> cd ~/1_project/data/untrimmed_fastq/</span></code></pre>
</div>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise-1"></a>
</h3>
<div class="callout-content">
<p>How big are the files? (Hint: Look at the options for the
<code>ls</code> command to see how to show file sizes.)</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-l</span> <span class="at">-h</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>-rw-rw-r-- 1 alice.palmer001 alice.palmer001 133M Aug  7 00:54 C1_S4_L001_R1_001_downsampled.fastq
-rw-rw-r-- 1 alice.palmer001 alice.palmer001 133M Aug  7 00:54 C1_S4_L001_R2_001_downsampled.fastq
-rw-rw-r-- 1 alice.palmer001 alice.palmer001 126M Aug  7 00:55 T1_S7_L001_R1_001_downsampled.fastq
-rw-rw-r-- 1 alice.palmer001 alice.palmer001 126M Aug  7 00:55 T1_S7_L001_R2_001_downsampled.fastq
-rw-rw-r-- 1 alice.palmer001 alice.palmer001 146M Aug  7 00:55 V1_S1_L001_R1_001_downsampled.fastq
-rw-rw-r-- 1 alice.palmer001 alice.palmer001 146M Aug  7 00:55 V1_S1_L001_R2_001_downsampled.fastq
</code></pre>
</div>
<p>There are six FASTQ files ranging from 126M (1246B) to 1465M.</p>
</div>
</div>
</div>
</div>
<p>FastQC can accept multiple file names as input, and on both zipped
and unzipped files, so we can use the *.fastq wildcard to run FastQC on
all of the FASTQ files in this directory.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">$</span> fastqc <span class="pp">*</span>.fastq</span></code></pre>
</div>
<p>You will see an automatically updating output message telling you the
progress of the analysis. It will start like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Started analysis of C1_S4_L001_R1_001_downsampled.fastq
Approx 5% complete for C1_S4_L001_R1_001_downsampled.fastq
Approx 10% complete for C1_S4_L001_R1_001_downsampled.fastq
Approx 15% complete for C1_S4_L001_R1_001_downsampled.fastq
Approx 20% complete for C1_S4_L001_R1_001_downsampled.fastq
Approx 25% complete for C1_S4_L001_R1_001_downsampled.fastq
Approx 30% complete for C1_S4_L001_R1_001_downsampled.fastq
Approx 35% complete for C1_S4_L001_R1_001_downsampled.fastq
Approx 40% complete for C1_S4_L001_R1_001_downsampled.fastq
Approx 45% complete for C1_S4_L001_R1_001_downsampled.fastq
Approx 50% complete for C1_S4_L001_R1_001_downsampled.fastq</code></pre>
</div>
<p>In total, it should take about thirty seconds for FastQC to run on
all six of our FASTQ files. When the analysis completes, your prompt
will return. So your screen will look something like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Approx 60% complete for V1_S1_L001_R2_001_downsampled.fastq
Approx 65% complete for V1_S1_L001_R2_001_downsampled.fastq
Approx 70% complete for V1_S1_L001_R2_001_downsampled.fastq
Approx 75% complete for V1_S1_L001_R2_001_downsampled.fastq
Approx 80% complete for V1_S1_L001_R2_001_downsampled.fastq
Approx 85% complete for V1_S1_L001_R2_001_downsampled.fastq
Approx 90% complete for V1_S1_L001_R2_001_downsampled.fastq
Approx 95% complete for V1_S1_L001_R2_001_downsampled.fastq
Analysis complete for V1_S1_L001_R2_001_downsampled.fastq
$</code></pre>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise-2"></a>
</h3>
<div class="callout-content">
<p>We did this directly on the command line, and it worked pretty
quickly (in part because we only have six files that have been
downsampled to only 5% of the original files). Usually for a task like
this we will want to use a slurm script. Turn your fastqc command into a
script and run it using <code>sbatch</code>.</p>
<p>Hint: You’ll need to load the module within the script and include
<code>sbatch</code> options.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="ex">$</span> nano fastqc.sh</span></code></pre>
</div>
<pre><code>#!/bin/bash

#SBATCH --job-name=fastqc # you can give your job a name
#SBATCH --nodes 1 # the number of processors or tasks
#SBATCH --cpus-per-task=2
#SBATCH --account=itcga # our account
#SBATCH --reservation=ITCGA_AUG2024 # this gives us special access during the workshop
#SBATCH --time=1:00:00 # the maximum time for the job
#SBATCH --mem=4gb # the amount of RAM
#SBATCH --partition=itcga # the specific server in chimera we are using
#SBATCH --error=%x-%A.err   # a filename to save error messages into
#SBATCH --output=%x-%A.out  # a filename to save any printed output into

# module load
module load fastqc-0.11.9-gcc-10.2.0-osi6pqc

# run fastqc
fastqc *.fastq</code></pre>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">$</span> sbatch fastqc.sh</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The FastQC program has created several new files within our
<code>data/untrimmed_fastq/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">C1_S4_L001_R1_001_downsampled.fastq</span></span>
<span><span class="va">C1_S4_L001_R1_001_downsampled_fastqc.html</span></span>
<span><span class="va">C1_S4_L001_R1_001_downsampled_fastqc.zip</span></span>
<span><span class="va">C1_S4_L001_R2_001_downsampled.fastq</span></span>
<span><span class="va">C1_S4_L001_R2_001_downsampled_fastqc.html</span></span>
<span><span class="va">C1_S4_L001_R2_001_downsampled_fastqc.zip</span></span>
<span><span class="va">T1_S7_L001_R1_001_downsampled.fastq</span></span>
<span><span class="va">T1_S7_L001_R1_001_downsampled_fastqc.html</span></span>
<span><span class="va">T1_S7_L001_R1_001_downsampled_fastqc.zip</span></span>
<span><span class="va">T1_S7_L001_R2_001_downsampled.fastq</span></span>
<span><span class="va">T1_S7_L001_R2_001_downsampled_fastqc.html</span></span>
<span><span class="va">T1_S7_L001_R2_001_downsampled_fastqc.zip</span></span>
<span><span class="va">V1_S1_L001_R1_001_downsampled.fastq</span></span>
<span><span class="va">V1_S1_L001_R1_001_downsampled_fastqc.html</span></span>
<span><span class="va">V1_S1_L001_R1_001_downsampled_fastqc.zip</span></span>
<span><span class="va">V1_S1_L001_R2_001_downsampled.fastq</span></span>
<span><span class="va">V1_S1_L001_R2_001_downsampled_fastqc.html</span></span>
<span><span class="va">V1_S1_L001_R2_001_downsampled_fastqc.zip</span></span></code></pre>
</div>
<p>For each input FASTQ file, FastQC has created a <code>.zip</code>
file and a <code>.html</code> file. The <code>.zip</code> file extension
indicates that this is actually a compressed set of multiple output
files. We will be working with these output files soon. The
<code>.html</code> file is a stable webpage displaying the summary
report for each of our samples.</p>
<p>We want to keep our data files and our results files separate, so we
will move these output files into a new directory within our
<code>results/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> ~/1_project/results/fastqc_untrimmed_reads</span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.zip ~/1_project/results/fastqc_untrimmed_reads/</span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.html ~/1_project/results/fastqc_untrimmed_reads/</span></code></pre>
</div>
<p>Now we can navigate into this results directory and do some closer
inspection of our output files.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="ex">$</span> cd ~/1_project/results/fastqc_untrimmed_reads/</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="viewing-the-fastqc-results">Viewing the FastQC results<a class="anchor" aria-label="anchor" href="#viewing-the-fastqc-results"></a></h3>
<p>If we were working on our local computers, we would be able to look
at each of these HTML files by opening them in a web browser.</p>
<p>However, these files are currently sitting on our remote AWS
instance, where our local computer can not see them. And, since we are
only logging into the AWS instance via the command line - it does not
have any web browser setup to display these files either.</p>
<p>So the easiest way to look at these webpage summary reports will be
to transfer them to our local computers (i.e. your laptop).</p>
<p>To transfer a file from a remote server to our own machines, we will
use <code>scp</code>, which we learned yesterday in the Shell Genomics
lesson.</p>
<p>First we will make a new directory on our computer to store the HTML
files we are transferring. Let’s put it on our desktop for now. Open a
new tab in your terminal program (you can use the pull down menu at the
top of your screen or the Cmd+t keyboard shortcut) and type:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> ~/Desktop/fastqc_html</span></code></pre>
</div>
<p>Now we can transfer our HTML files to our local computer using
<code>scp</code>. Remember to replace the username with your own!</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="ex">$</span> scp your.UMB.username@chimera.umb.edu:/itcgastorage/share_home/your.UMB.username/itcga_workshop/untrimmed_fastq/fastqc_untrimmed_reads/<span class="pp">*</span>.html ~/Desktop/fastqc_html</span></code></pre>
</div>
<div id="note-on-using-zsh" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="note-on-using-zsh" class="callout-inner">
<h3 class="callout-title">Note on using zsh<a class="anchor" aria-label="anchor" href="#note-on-using-zsh"></a>
</h3>
<div class="callout-content">
<p>If you are using zsh instead of bash (macOS for example changed the
default recently to zsh), it is likely that a
<code>no matches found</code> error will be displayed. The reason for
this is that the wildcard (“*”) is not correctly interpreted. To fix
this problem the wildcard needs to be escaped with a “\”:</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="ex">$</span> scp your.UMB.username@chimera.umb.edu:/itcgastorage/share_home/your.UMB.username/itcga_workshop/untrimmed_fastq/fastqc_untrimmed_reads/\*.html ~/Desktop/fastqc_html</span></code></pre>
</div>
<p>Alternatively, you can put the whole path into quotation marks:</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="ex">$</span> scp <span class="st">"your.UMB.username@chimera.umb.edu:/itcgastorage/share_home/your.UMB.username/itcga_workshop/untrimmed_fastq/fastqc_untrimmed_reads/*.html"</span> ~/Desktop/fastqc_html</span></code></pre>
</div>
</div>
</div>
</div>
<p>The second part starts with a <code>:</code> and then gives the
absolute path of the files you want to transfer from your remote
computer. Do not forget the <code>:</code>. We used a wildcard
(<code>*.html</code>) to indicate that we want all of the HTML
files.</p>
<p>The third part of the command gives the absolute path of the location
you want to put the files. This is on your local computer and is the
directory we just created <code>~/Desktop/fastqc_html</code>.</p>
<p>You should see a status output like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>your.UMB.username@chimera.umb.edu's password:
C1_S4_L001_R1_001_downsampled_fastqc.html     100%  626KB   7.4MB/s   00:00
C1_S4_L001_R2_001_downsampled_fastqc.html     100%  627KB   9.1MB/s   00:00
T1_S7_L001_R1_001_downsampled_fastqc.html     100%  629KB  10.9MB/s   00:00
T1_S7_L001_R2_001_downsampled_fastqc.html     100%  626KB   8.4MB/s   00:00
V1_S1_L001_R1_001_downsampled_fastqc.html     100%  627KB  10.3MB/s   00:00
V1_S1_L001_R2_001_downsampled_fastqc.html     100%  630KB  14.5MB/s   00:00 </code></pre>
</div>
<p>Now we can go to our new directory and open the 6 HTML files.</p>
<p>Depending on your system, you should be able to select and open them
all at once via a right click menu in your file browser.</p>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise-3"></a>
</h3>
<div class="callout-content">
<p>Discuss your results with a neighbor. Which sample(s) looks the best
in terms of per base sequence quality? Which sample(s) look the
worst?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>All of the reads contain usable data, but the quality decreases
toward the beginning and end of the reads.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="decoding-the-other-fastqc-outputs">Decoding the other FastQC outputs<a class="anchor" aria-label="anchor" href="#decoding-the-other-fastqc-outputs"></a></h3>
<p>We have now looked at quite a few “Per base sequence quality” FastQC
graphs, but there are nine other graphs that we have not talked about!
Below we have provided a brief overview of interpretations for each of
these plots. For more information, please see the <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/" class="external-link">FastQC
documentation here</a></p>
<ul><li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/12%20Per%20Tile%20Sequence%20Quality.html" class="external-link"><strong>Per
tile sequence quality</strong></a>: the machines that perform sequencing
are divided into tiles. This plot displays patterns in base quality
along these tiles. Consistently low scores are often found around the
edges, but hot spots can also occur in the middle if an air bubble was
introduced at some point during the run.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/3%20Per%20Sequence%20Quality%20Scores.html" class="external-link"><strong>Per
sequence quality scores</strong></a>: a density plot of quality for all
reads at all positions. This plot shows what quality scores are most
common.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/4%20Per%20Base%20Sequence%20Content.html" class="external-link"><strong>Per
base sequence content</strong></a>: plots the proportion of each base
position over all of the reads. Typically, we expect to see each base
roughly 25% of the time at each position, but this often fails at the
beginning or end of the read due to quality or adapter content.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/5%20Per%20Sequence%20GC%20Content.html" class="external-link"><strong>Per
sequence GC content</strong></a>: a density plot of average GC content
in each of the reads.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/6%20Per%20Base%20N%20Content.html" class="external-link"><strong>Per
base N content</strong></a>: the percent of times that ‘N’ occurs at a
position in all reads. If there is an increase at a particular position,
this might indicate that something went wrong during sequencing.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/7%20Sequence%20Length%20Distribution.html" class="external-link"><strong>Sequence
Length Distribution</strong></a>: the distribution of sequence lengths
of all reads in the file. If the data is raw, there is often on sharp
peak, however if the reads have been trimmed, there may be a
distribution of shorter lengths.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/8%20Duplicate%20Sequences.html" class="external-link"><strong>Sequence
Duplication Levels</strong></a>: A distribution of duplicated sequences.
In sequencing, we expect most reads to only occur once. If some
sequences are occurring more than once, it might indicate enrichment
bias (e.g. from PCR). If the samples are high coverage (or RNA-seq or
amplicon), this might not be true.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/9%20Overrepresented%20Sequences.html" class="external-link"><strong>Overrepresented
sequences</strong></a>: A list of sequences that occur more frequently
than would be expected by chance.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/10%20Adapter%20Content.html" class="external-link"><strong>Adapter
Content</strong></a>: a graph indicating where adapater sequences occur
in the reads.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/11%20Kmer%20Content.html" class="external-link"><strong>K-mer
Content</strong></a>: a graph showing any sequences which may show a
positional bias within the reads.</li>
</ul></div>
<div class="section level3">
<h3 id="working-with-the-fastqc-text-output">Working with the FastQC text output<a class="anchor" aria-label="anchor" href="#working-with-the-fastqc-text-output"></a></h3>
<p>Now that we have looked at our HTML reports to get a feel for the
data, let’s look more closely at the other output files. Go back to the
tab in your terminal program that is connected to the chimera (the tab
label will start with <code>your.UMB.username@chimerahead</code>) and
make sure you are in our results subdirectory.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="ex">$</span> cd ~/1_project/results/fastqc_untrimmed_reads/</span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">C1_S4_L001_R1_001_downsampled_fastqc.html</span></span>
<span><span class="va">C1_S4_L001_R1_001_downsampled_fastqc.zip</span></span>
<span><span class="va">C1_S4_L001_R2_001_downsampled_fastqc.html</span></span>
<span><span class="va">C1_S4_L001_R2_001_downsampled_fastqc.zip</span></span>
<span><span class="va">T1_S7_L001_R1_001_downsampled_fastqc.html</span></span>
<span><span class="va">T1_S7_L001_R1_001_downsampled_fastqc.zip</span></span>
<span><span class="va">T1_S7_L001_R2_001_downsampled_fastqc.html</span></span>
<span><span class="va">T1_S7_L001_R2_001_downsampled_fastqc.zip</span></span>
<span><span class="va">V1_S1_L001_R1_001_downsampled_fastqc.html</span></span>
<span><span class="va">V1_S1_L001_R1_001_downsampled_fastqc.zip</span></span>
<span><span class="va">V1_S1_L001_R2_001_downsampled_fastqc.html</span></span>
<span><span class="va">V1_S1_L001_R2_001_downsampled_fastqc.zip</span></span></code></pre>
</div>
<p>Our <code>.zip</code> files are compressed files. They each contain
multiple different types of output files for a single input FASTQ file.
To view the contents of a <code>.zip</code> file, we can use the program
<code>unzip</code> to decompress these files. Let’s try doing them all
at once using a wildcard.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="ex">$</span> unzip <span class="pp">*</span>.zip</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Archive:  C1_S4_L001_R1_001_downsampled_fastqc.zip
caution: filename not matched:  C1_S4_L001_R2_001_downsampled_fastqc.zip
caution: filename not matched:  T1_S7_L001_R1_001_downsampled_fastqc.zip
caution: filename not matched:  T1_S7_L001_R2_001_downsampled_fastqc.zip
caution: filename not matched:  V1_S1_L001_R1_001_downsampled_fastqc.zip
caution: filename not matched:  V1_S1_L001_R2_001_downsampled_fastqc.zip</code></pre>
</div>
<p>This did not work. We unzipped the first file and then got a warning
message for each of the other <code>.zip</code> files. This is because
<code>unzip</code> expects to get only one zip file as input. We could
go through and unzip each file one at a time, but this is very time
consuming and error-prone. Someday you may have 500 files to unzip!</p>
<p>A more efficient way is to use a <code>for</code> loop like we
learned in the Shell Genomics lesson to iterate through all of our
<code>.zip</code> files. Let’s see what that looks like and then we will
discuss what we are doing with each line of our loop.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="ex">$</span> for filename in <span class="pp">*</span>.zip</span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a><span class="op">&gt;</span> do</span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a><span class="op">&gt;</span> unzip <span class="va">$filename</span></span>
<span id="cb34-4"><a href="#cb34-4" tabindex="-1"></a><span class="op">&gt;</span> done</span></code></pre>
</div>
<p>In this example, the input is six filenames (one filename for each of
our <code>.zip</code> files). Each time the loop iterates, it will
assign a file name to the variable <code>filename</code> and run the
<code>unzip</code> command. The first time through the loop,
<code>$filename</code> is
<code>C1_S4_L001_R1_001_downsampled_fastqc.zip</code>. The interpreter
runs the command <code>unzip</code> on
<code>C1_S4_L001_R1_001_downsampled_fastqc.zip</code>. For the second
iteration, <code>$filename</code> becomes
<code>C1_S4_L001_R2_001_downsampled_fastqc.zip</code>. This time, the
shell runs <code>unzip</code> on
<code>C1_S4_L001_R2_001_downsampled_fastqc.zip</code>. It then repeats
this process for the four other <code>.zip</code> files in our
directory.</p>
<p>When we run our <code>for</code> loop, you will see output that
starts like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">Archive</span><span class="op">:</span>  <span class="va">C1_S4_L001_R1_001_downsampled_fastqc.zip</span></span>
<span>   <span class="va">creating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span></span>
<span>   <span class="va">creating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span></span>
<span>   <span class="va">creating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">fastqc_icon.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">warning.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">error.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">tick.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">summary.txt</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_quality.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_tile_quality.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_sequence_quality.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_sequence_content.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_sequence_gc_content.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_n_content.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">sequence_length_distribution.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">duplication_levels.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">adapter_content.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">fastqc_report.html</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">fastqc_data.txt</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span><span class="va">fastqc.fo</span></span>
<span><span class="va">Archive</span><span class="op">:</span>  <span class="va">C1_S4_L001_R2_001_downsampled_fastqc.zip</span></span>
<span>   <span class="va">creating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span></span>
<span>   <span class="va">creating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span></span>
<span>   <span class="va">creating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">fastqc_icon.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">warning.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">error.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">tick.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">summary.txt</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_quality.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_tile_quality.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_sequence_quality.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_sequence_content.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_sequence_gc_content.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_n_content.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">sequence_length_distribution.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">duplication_levels.png</span> </span></code></pre>
</div>
<p>The <code>unzip</code> program is decompressing the <code>.zip</code>
files and creating a new directory (with subdirectories) for each of our
samples, to store all of the different output that is produced by
FastQC. There</p>
<p>are a lot of files here. The one we are going to focus on is the
<code>summary.txt</code> file.</p>
<p>If you list the files in our directory now you will see:</p>
<pre><code><span><span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span></span>
<span><span class="va">C1_S4_L001_R1_001_downsampled_fastqc.html</span></span>
<span><span class="va">C1_S4_L001_R1_001_downsampled_fastqc.zip</span></span>
<span><span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span></span>
<span><span class="va">C1_S4_L001_R2_001_downsampled_fastqc.html</span></span>
<span><span class="va">C1_S4_L001_R2_001_downsampled_fastqc.zip</span></span>
<span><span class="va">T1_S7_L001_R1_001_downsampled_fastqc</span></span>
<span><span class="va">T1_S7_L001_R1_001_downsampled_fastqc.html</span></span>
<span><span class="va">T1_S7_L001_R1_001_downsampled_fastqc.zip</span></span>
<span><span class="va">T1_S7_L001_R2_001_downsampled_fastqc</span></span>
<span><span class="va">T1_S7_L001_R2_001_downsampled_fastqc.html</span></span>
<span><span class="va">T1_S7_L001_R2_001_downsampled_fastqc.zip</span></span>
<span><span class="va">V1_S1_L001_R1_001_downsampled_fastqc</span></span>
<span><span class="va">V1_S1_L001_R1_001_downsampled_fastqc.html</span></span>
<span><span class="va">V1_S1_L001_R1_001_downsampled_fastqc.zip</span></span>
<span><span class="va">V1_S1_L001_R2_001_downsampled_fastqc</span></span>
<span><span class="va">V1_S1_L001_R2_001_downsampled_fastqc.html</span></span>
<span><span class="va">V1_S1_L001_R2_001_downsampled_fastqc.zip</span></span></code></pre>
<p>{:. output}</p>
<p>The <code>.html</code> files and the compressed <code>.zip</code>
files are still present, but now we also have a new directory for each
of our samples. We can see for sure that it is a directory if we use the
<code>-F</code> flag for <code>ls</code>.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-F</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">C1_S4_L001_R1_001_downsampled_fastqc</span><span class="op">/</span></span>
<span><span class="va">C1_S4_L001_R1_001_downsampled_fastqc.html</span></span>
<span><span class="va">C1_S4_L001_R1_001_downsampled_fastqc.zip</span></span>
<span><span class="va">C1_S4_L001_R2_001_downsampled_fastqc</span><span class="op">/</span></span>
<span><span class="va">C1_S4_L001_R2_001_downsampled_fastqc.html</span></span>
<span><span class="va">C1_S4_L001_R2_001_downsampled_fastqc.zip</span></span>
<span><span class="va">T1_S7_L001_R1_001_downsampled_fastqc</span><span class="op">/</span></span>
<span><span class="va">T1_S7_L001_R1_001_downsampled_fastqc.html</span></span>
<span><span class="va">T1_S7_L001_R1_001_downsampled_fastqc.zip</span></span>
<span><span class="va">T1_S7_L001_R2_001_downsampled_fastqc</span><span class="op">/</span></span>
<span><span class="va">T1_S7_L001_R2_001_downsampled_fastqc.html</span></span>
<span><span class="va">T1_S7_L001_R2_001_downsampled_fastqc.zip</span></span>
<span><span class="va">V1_S1_L001_R1_001_downsampled_fastqc</span><span class="op">/</span></span>
<span><span class="va">V1_S1_L001_R1_001_downsampled_fastqc.html</span></span>
<span><span class="va">V1_S1_L001_R1_001_downsampled_fastqc.zip</span></span>
<span><span class="va">V1_S1_L001_R2_001_downsampled_fastqc</span><span class="op">/</span></span>
<span><span class="va">V1_S1_L001_R2_001_downsampled_fastqc.html</span></span>
<span><span class="va">V1_S1_L001_R2_001_downsampled_fastqc.zip</span></span></code></pre>
</div>
<p>Let’s see what files are present within one of these output
directories.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-F</span> C1_S4_L001_R1_001_downsampled_fastqc/</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>fastqc_data.txt  fastqc.fo  fastqc_report.html	Icons/	Images/  summary.txt</code></pre>
</div>
<p>Use <code>less</code> to preview the <code>summary.txt</code> file
for this sample.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a><span class="ex">$</span> less C1_S4_L001_R1_001_downsampled_fastqc/summary.txt</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>PASS    Basic Statistics        C1_S4_L001_R1_001_downsampled.fastq
PASS    Per base sequence quality       C1_S4_L001_R1_001_downsampled.fastq
PASS    Per tile sequence quality       C1_S4_L001_R1_001_downsampled.fastq
PASS    Per sequence quality scores     C1_S4_L001_R1_001_downsampled.fastq
FAIL    Per base sequence content       C1_S4_L001_R1_001_downsampled.fastq
WARN    Per sequence GC content C1_S4_L001_R1_001_downsampled.fastq
PASS    Per base N content      C1_S4_L001_R1_001_downsampled.fastq
PASS    Sequence Length Distribution    C1_S4_L001_R1_001_downsampled.fastq
FAIL    Sequence Duplication Levels     C1_S4_L001_R1_001_downsampled.fastq
WARN    Overrepresented sequences       C1_S4_L001_R1_001_downsampled.fastq
PASS    Adapter Content C1_S4_L001_R1_001_downsampled.fastq</code></pre>
</div>
<p>The summary file gives us a list of tests that FastQC ran, and tells
us whether this sample passed, failed, or is borderline
(<code>WARN</code>). Remember, to quit from <code>less</code> you must
type <code>q</code>.</p>
</div>
<div class="section level3">
<h3 id="documenting-our-work">Documenting our work<a class="anchor" aria-label="anchor" href="#documenting-our-work"></a></h3>
<p>We can make a record of the results we obtained for all our samples
by concatenating all of our <code>summary.txt</code> files into a single
file using the <code>cat</code> command. We will call this
<code>fastqc_summaries.txt</code> and move it to
<code>~/1_project/docs</code>.</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a><span class="ex">$</span> cat <span class="pp">*</span>/summary.txt <span class="op">&gt;</span> ~/1_project/docs/fastqc_summaries.txt</span></code></pre>
</div>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise-4"></a>
</h3>
<div class="callout-content">
<p>Which samples failed at least one of FastQC’s quality tests? What
test(s) did those samples fail?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<p>We can get the list of all failed tests using <code>grep</code>.</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="ex">$</span> cd ~/1_project/docs</span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a><span class="ex">$</span> grep FAIL fastqc_summaries.txt</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>FAIL	Per base sequence content	C1_S4_L001_R1_001_downsampled.fastq
FAIL	Sequence Duplication Levels	C1_S4_L001_R1_001_downsampled.fastq
FAIL	Per base sequence content	C1_S4_L001_R2_001_downsampled.fastq
FAIL	Sequence Duplication Levels	C1_S4_L001_R2_001_downsampled.fastq
FAIL	Per base sequence content	T1_S7_L001_R1_001_downsampled.fastq
FAIL	Sequence Duplication Levels	T1_S7_L001_R1_001_downsampled.fastq
FAIL	Per base sequence content	T1_S7_L001_R2_001_downsampled.fastq
FAIL	Sequence Duplication Levels	T1_S7_L001_R2_001_downsampled.fastq
FAIL	Per base sequence content	V1_S1_L001_R1_001_downsampled.fastq
FAIL	Sequence Duplication Levels	V1_S1_L001_R1_001_downsampled.fastq
FAIL	Per base sequence content	V1_S1_L001_R2_001_downsampled.fastq
FAIL	Per sequence GC content	V1_S1_L001_R2_001_downsampled.fastq
FAIL	Sequence Duplication Levels	V1_S1_L001_R2_001_downsampled.fastq
</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="other-notes-optional">Other notes – optional<a class="anchor" aria-label="anchor" href="#other-notes-optional"></a>
<a class="anchor" aria-label="anchor" href="#other-notes-optional"></a></h2>
<hr class="half-width"><div id="quality-encodings-vary" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="quality-encodings-vary" class="callout-inner">
<h3 class="callout-title">Quality encodings vary<a class="anchor" aria-label="anchor" href="#quality-encodings-vary"></a>
</h3>
<div class="callout-content">
<p>Although we have used a particular quality encoding system to
demonstrate interpretation of read quality, different sequencing
machines use different encoding systems. This means that, depending on
which sequencer you use to generate your data, a <code>#</code> may not
be an indicator of a poor quality base call.</p>
<p>This mainly relates to older Solexa/Illumina data, but it is
essential that you know which sequencing platform was used to generate
your data, so that you can tell your quality control program which
encoding to use. If you choose the wrong encoding, you run the risk of
throwing away good reads or (even worse) not throwing away bad
reads!</p>
</div>
</div>
</div>
<div id="same-symbols-different-meanings" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="same-symbols-different-meanings" class="callout-inner">
<h3 class="callout-title">Same symbols, different meanings<a class="anchor" aria-label="anchor" href="#same-symbols-different-meanings"></a>
</h3>
<div class="callout-content">
<p>Here we see <code>&gt;</code> being used as a shell prompt, whereas
<code>&gt;</code> is also used to redirect output. Similarly,
<code>$</code> is used as a shell prompt, but, as we saw earlier, it is
also used to ask the shell to get the value of a variable.</p>
<p>If the <em>shell</em> prints <code>&gt;</code> or <code>$</code> then
it expects you to type something, and the symbol is a prompt.</p>
<p>If <em>you</em> type <code>&gt;</code> or <code>$</code> yourself, it
is an instruction from you that the shell should redirect output or get
the value of a variable.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>Quality encodings vary across sequencing platforms.</li>
<li>
<code>for</code> loops let you perform the same set of operations on
multiple files with a single command.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="01-background.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="03-trimming.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="01-background.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction and
        </a>
        <a class="chapter-link float-end" href="03-trimming.html" rel="next">
          Next: Trimming and...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/btmoyers/wrangling-transcriptomics/edit/main/episodes/02-quality-control.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/btmoyers/wrangling-transcriptomics/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/btmoyers/wrangling-transcriptomics/" class="external-link">Source</a></p>
				<p><a href="https://github.com/btmoyers/wrangling-transcriptomics/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:brook.moyers@gmail.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.5" class="external-link">sandpaper (0.16.5)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.6" class="external-link">pegboard (0.7.6)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.3" class="external-link">varnish (1.0.3)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://btmoyers.github.io/wrangling-transcriptomics/02-quality-control.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Assessing Read Quality",
  "creativeWorkStatus": "active",
  "url": "https://btmoyers.github.io/wrangling-transcriptomics/02-quality-control.html",
  "identifier": "https://btmoyers.github.io/wrangling-transcriptomics/02-quality-control.html",
  "dateCreated": "2024-08-06",
  "dateModified": "2024-08-19",
  "datePublished": "2024-08-19"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

