<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Data Wrangling and Processing for Transcriptomics: RNA-Seq Workflow</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png"><link rel="manifest" href="favicons/incubator/site.webmanifest"><link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/04-variant_calling.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Data Wrangling and Processing for Transcriptomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Data Wrangling and Processing for Transcriptomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Data Wrangling and Processing for Transcriptomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 62%" class="percentage">
    62%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 62%" aria-valuenow="62" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/04-variant_calling.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-background.html">1. Introduction and Experimental Planning Considerations for RNA-Seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-quality-control.html">2. Assessing Read Quality</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-trimming.html">3. Trimming and Filtering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        4. RNA-Seq Workflow
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#alignment-to-a-reference-genome">Alignment to a reference genome</a></li>
<li><a href="#feature-read-counting">Feature (read) counting</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-automation.html">5. Analyzing Read Count Data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li><li><a href="reference.html">Glossary</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="03-trimming.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="05-automation.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="03-trimming.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Trimming and
        </a>
        <a class="chapter-link float-end" href="05-automation.html" rel="next">
          Next: Analyzing Read Count...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>RNA-Seq Workflow</h1>
        <p>Last updated on 2024-08-21 |

        <a href="https://github.com/btmoyers/wrangling-transcriptomics/edit/main/episodes/04-variant_calling.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How do I find gene expression differences between my samples?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Understand the steps involved in aligning fastq files and extracting
relevant information from the alignments.</li>
<li>Describe the types of data formats encountered during this
workflow.</li>
<li>Use command line tools to perform alignment and read counting.</li>
</ul></div>
</div>
</div>
</div>
</div>
<p>We mentioned before that we are working with files from a study of
fibroblast to myofibroblast phenoconversion, <a href="https://www.doi.org/10.1038/s41598-018-21506-7" class="external-link">Patalano et
al. 2018</a>. Now that we have looked at our data to make sure that it
is high quality, and removed low-quality base calls, we can perform
alignment and read counting. We care how the three treatments (C =
CXCL12, T = TFG-beta, and V = Vehicle control) compare to each other, to
understand the roles of each axis in the biological process of
phenoconversion.</p>
<section><h2 class="section-heading" id="alignment-to-a-reference-genome">Alignment to a reference genome<a class="anchor" aria-label="anchor" href="#alignment-to-a-reference-genome"></a></h2>
<hr class="half-width"><figure><img src="fig/variant_calling_workflow_align.png" alt="workflow_align" class="figure mx-auto d-block"></figure><p>We perform read alignment or mapping to determine where in the genome
our reads originated from. There are a number of tools to choose from
and, while there is no gold standard, there are some tools that are
better suited for particular analyses. We will be using <a href="https://daehwankimlab.github.io/hisat2/" class="external-link">HISAT2</a>, which is a
fast and sensitive alignment program for mapping next-generation
sequencing reads (both DNA and RNA) to a population of human genomes as
well as to a single reference genome.</p>
<p>The alignment process consists of two steps:</p>
<ol style="list-style-type: decimal"><li>Indexing the reference genome</li>
<li>Aligning the reads to the reference genome</li>
</ol><div class="section level3">
<h3 id="setting-up">Setting up<a class="anchor" aria-label="anchor" href="#setting-up"></a></h3>
<p>First we download or access a reference genome for <em>Homo
sapiens</em>, <a href="https://grch37.ensembl.org/Homo_sapiens/Info/Index" class="external-link">GRCh37</a>. if
you remember, we earlier downloaded a README file from the Ensembl
database about this genome. We could download the files we need using
<code>curl</code> or <code>wget</code> from the database, but it will
take a long time since the files are big! Instead, we have already
downloaded them into a shared directory, and we can use <code>cp</code>
or, to save on storage space, <a href="https://kb.iu.edu/d/abbe" class="external-link">soft
linking</a> to access them.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> cd ~/1_project</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> data/genome</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="ex">$</span> cd data/genome/</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="ex">$</span> ln <span class="at">-s</span> /itcgastorage/data01/itcga_workshops/aug2024_genomics/Genome/hg38/Homo_sapiens.GRCh38.dna.primary_assembly.fa</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">Homo_sapiens.GRCh38.dna.primary_assembly.fa</span></span></code></pre>
</div>
<p>Here, <code>ln -s</code> creates a symbolic link between the first
file and the new name (or location) you have given it. This is really
useful when using a shared set of files that you don’t want to copy or
accidentally mess up. Note that you can give it a new name or a new
path, but if you don’t it will create a link to the specified file in
your current working directory, as it did above.</p>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>We also want to be able to access the gtf, exons, and ss files stored
in the same shared directory
<code>/itcgastorage/data01/itcga_workshops/aug2024_genomics/Genome/hg38/</code>.
Make soft links in your <code>genome</code> directory for all three
files as well.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">$</span> ln <span class="at">-s</span> /itcgastorage/data01/itcga_workshops/aug2024_genomics/Genome/hg38/Homo_sapiens.GRCh38.111.gtf</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="ex">$</span> ln <span class="at">-s</span> /itcgastorage/data01/itcga_workshops/aug2024_genomics/Genome/hg38/Homo_sapiens.GRCh38_exons_file.txt</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="ex">$</span> ln <span class="at">-s</span> /itcgastorage/data01/itcga_workshops/aug2024_genomics/Genome/hg38/Homo_sapiens.GRCh38_ss_file.txt</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="index-the-reference-genome">Index the reference genome<a class="anchor" aria-label="anchor" href="#index-the-reference-genome"></a></h3>
<p>Our first step is to index the reference genome for use by HISAT2.
Indexing allows the aligner to quickly find potential alignment sites
for query sequences in a genome, which saves time during alignment.
Indexing the reference only has to be run once. The only reason you
would want to create a new index is if you are working with a different
reference genome or you are using a different tool for alignment. Make a
slurm script to run the indexing, making sure to include the following
SBATCH options and to load the following modules before the hisat2-build
command:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co">#SBATCH --job-name=gindex # you can give your job a name</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#SBATCH --ntasks=24 # the number of processors or tasks</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#SBATCH --account=itcga # our account</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#SBATCH --reservation=ITCGA_AUG2024 # this gives us special access during the workshop</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#SBATCH --time=10:00:00 # the maximum time for the job</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#SBATCH --mem=32gb # the amount of RAM</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#SBATCH --partition=itcga # the specific server in chimera we are using</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#SBATCH --error=%x-%A.err   # a filename to save error messages into</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#SBATCH --output=%x-%A.out  # a filename to save any printed output into</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="ex">module</span> load gcc-10.2.0-gcc-9.3.0-f3oaqv7</span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="ex">module</span> load python-3.8.12-gcc-10.2.0-oe4tgov</span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="ex">module</span> load hisat2-2.1.0-gcc-9.3.0-u7zbyow</span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a><span class="ex">hisat2-build</span> <span class="at">-p</span> 24 ~/1_project/data/genome/Homo_sapiens.GRCh38.dna.primary_assembly.fa <span class="at">--ss</span> ~/1_project/data/genome/Homo_sapiens.GRCh38_ss_file.txt <span class="at">--exon</span> ~/1_project/data/genome/Homo_sapiens.GRCh38_exons_file.txt ~/1_project/data/genome/hg38</span></code></pre>
</div>
<p>Don’t forget the <code>#!/bin/bash/</code> at the top of your script
before you run the job! You could also make the script more generalized
by making the genome path a variable given on the command line.</p>
<p>While the index is created, you would see output that looks something
like this if you peek in the .err log file your job is creating.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Settings:
  Output files: "..*.ht2"
  Line rate: 7 (line is 128 bytes)
  Lines per side: 1 (side is 128 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Local offset rate: 3 (one in 8)
  Local fTable chars: 6
  Local sequence length: 57344
  Local sequence overlap between two consecutive indexes: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  Homo_sapiens.GRCh38.dna.primary_assembly.fa</code></pre>
</div>
<p>Indexing can take quite a while, but we wanted to give you the tools
to do this if you end up working with a different genome. For not, we
can also access the pre-indexed genome in the same directory that we
soft-linked from,
<code>/itcgastorage/data01/itcga_workshops/aug2024_genomics/Genome/hg38/</code>.</p>
<p>Before aligning, let’s make a few directories to store the results of
our next few steps.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">$</span> cd ~/1_project</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> results/sam results/bam results/counts</span></code></pre>
</div>
<div class="section level4">
<h4 id="align-reads-to-reference-genome">Align reads to reference genome<a class="anchor" aria-label="anchor" href="#align-reads-to-reference-genome"></a></h4>
<p>The alignment process consists of choosing an appropriate reference
genome to map our reads against and then deciding on an aligner. We will
use the HISAT2 program because it performs well with human RNA-Seq
data.</p>
<p>An example of what a <code>hisat2</code> command looks like is below.
This command will not run, as we do not have the files
<code>genome.fa</code>, <code>input_file_R1.fastq</code>, or
<code>input_file_R2.fastq</code>.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">$</span> hisat2 <span class="at">-p</span> 24 <span class="at">-x</span> /itcgastorage/data01/itcga_workshops/aug2024_genomics/Genome/hg38/hg38 <span class="at">-1</span> input_file_R1.fastq <span class="at">-2</span> input_file_R2.fastq <span class="at">-S</span> ~/1_project/results/sam/output_file.sam</span></code></pre>
</div>
<p>On chimera, we’ll need to load these modules (same as we did for
indexing):</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">$</span> module load gcc-10.2.0-gcc-9.3.0-f3oaqv7</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="ex">$</span> module load python-3.8.12-gcc-10.2.0-oe4tgov</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="ex">$</span> module load hisat2-2.1.0-gcc-9.3.0-u7zbyow</span></code></pre>
</div>
<p>Have a look at the <a href="https://daehwankimlab.github.io/hisat2/manual/" class="external-link">HISAT2 options
page</a>. While we are running HISAT2 with the default parameters here,
different uses might require a change of parameters. <em>NOTE: Always
read the manual page for any tool before using and make sure the options
you use are appropriate for your data.</em></p>
<p>We are going to start by aligning the reads from just one of the
samples in our dataset (the trimmed pair of <code>V1_S1_L001</code>
FASTQ files). Later, we will be iterating this whole process on all of
our sample files.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="ex">$</span> hisat2 <span class="at">-p</span> 24 <span class="at">-x</span> /itcgastorage/data01/itcga_workshops/aug2024_genomics/Genome/hg38/hg38 <span class="at">-1</span> ~/1_project/data/trim_fastq/V1_S1_L001_R1_001_ds_trim.fastq <span class="at">-2</span> ~/1_project/data/trim_fastq/V1_S1_L001_R2_001_ds_trim.fastq <span class="at">-S</span> ~/1_project/results/sam/V1_S1_L001_ds_trim_align.sam</span></code></pre>
</div>
<p>You will see output that looks like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>961680 reads; of these:
  961680 (100.00%) were paired; of these:
    450677 (46.86%) aligned concordantly 0 times
    440285 (45.78%) aligned concordantly exactly 1 time
    70718 (7.35%) aligned concordantly &gt;1 times
    ----
    450677 pairs aligned concordantly 0 times; of these:
      4041 (0.90%) aligned discordantly 1 time
    ----
    446636 pairs aligned 0 times concordantly or discordantly; of these:
      893272 mates make up the pairs; of these:
        857773 (96.03%) aligned 0 times
        25122 (2.81%) aligned exactly 1 time
        10377 (1.16%) aligned &gt;1 times
55.40% overall alignment rate</code></pre>
</div>
<p>Overall, a bit over half of the reads in this sample aligned to the
reference genome we are using (overall alignment rate), and most of
these aligned to a single location, the same place in the genome as
their pair (“aligned concordantly exactly one time”). A small proportion
of reads did something else: aligned in a pair in more than one genome
position (“aligned concordantly &gt;1 time”), aligned but each pair was
in a different location (“aligned discordantly”), or only one of the
pair aligned (“pairs aligned 0 times”, “mates… aligned”). This isn’t
great for a sequence file that has gone through quality control, but it
is still workable for us!</p>
<div class="section level5">
<h5 id="sambam-format">SAM/BAM format<a class="anchor" aria-label="anchor" href="#sambam-format"></a></h5>
<p>The <a href="https://genome.sph.umich.edu/wiki/SAM" class="external-link">SAM file</a>, is
a tab-delimited text file that contains information for each individual
read and its alignment to the genome. While we do not have time to go
into detail about the features of the SAM format, the paper by <a href="https://bioinformatics.oxfordjournals.org/content/25/16/2078.full" class="external-link">Heng
Li et al.</a> provides a lot more detail on the specification.</p>
<p><strong>The compressed binary version of SAM is called a BAM
file.</strong> We use this version to reduce size and to allow for
<em>indexing</em>, which enables efficient random access of the data
contained within the file.</p>
<p>The file begins with a <strong>header</strong>, which is optional.
The header is used to describe the source of data, reference sequence,
method of alignment, etc., this will change depending on the aligner
being used. Following the header is the <strong>alignment
section</strong>. Each line that follows corresponds to alignment
information for a single read. Each alignment line has <strong>11
mandatory fields</strong> for essential mapping information and a
variable number of other fields for aligner specific information. An
example entry from a SAM file is displayed below with the different
fields highlighted.</p>
<figure><img src="fig/sam_bam.png" alt="sam_bam1" class="figure mx-auto d-block"></figure><figure><img src="fig/sam_bam3.png" alt="sam_bam2" class="figure mx-auto d-block"></figure><p>We will convert the SAM file to BAM format using the
<code>samtools</code> program with the <code>view</code> command (yeah,
it’s a funny name for the conversion process) and tell this command that
the input is in SAM format (<code>-S</code>) and to output BAM format
(<code>-b</code>). First we’ll need to load the samtools module.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">$</span> module load samtools-1.10-gcc-9.3.0-flukja5</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="ex">$</span> cd ~/1_project</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="ex">$</span> samtools view <span class="at">-S</span> <span class="at">-b</span> results/sam/V1_S1_L001_ds_trim_align.sam <span class="op">&gt;</span> results/bam/V1_S1_L001_ds_trim_align.bam</span></code></pre>
</div>
<p>This process can take a little while! You can speed it up using the
<code>-@</code> option to allocate additional CPUs to the task (e.g.,
try <code>-@ 4</code>).</p>
</div>
</div>
<div class="section level4">
<h4 id="sort-bam-file-by-coordinates">Sort BAM file by coordinates<a class="anchor" aria-label="anchor" href="#sort-bam-file-by-coordinates"></a></h4>
<p>Next we sort the BAM file using the <code>sort</code> command from
<code>samtools</code>. <code>-o</code> tells the command where to write
the output. Notice that we are saving the file back into the same name,
so the previous unsorted file will be overwritten by the sorted one.
This saves us some space, but might be a bad idea if want to know which
files are sorted and which are not. In that case, you would give the
output a different name.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">$</span> samtools sort <span class="at">-o</span> results/bam/V1_S1_L001_ds_trim_align.bam results/bam/V1_S1_L001_ds_trim_align.bam</span></code></pre>
</div>
<p>Our files are pretty small, so we will not see this output. If you
run the workflow with larger files, you might see something like
this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[bam_sort_core] merging from 2 files...</code></pre>
</div>
<p>SAM/BAM files can be sorted in multiple ways, e.g. by location of
alignment on the chromosome, by read name, etc. It is important to be
aware that different alignment tools will output differently sorted
SAM/BAM, and different downstream tools require differently sorted
alignment files as input. The default, which we are using here, is to
sort by the alignment coordinates in the reference genome.</p>
<p>You can use samtools to learn more about this sorted bam file as
well.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">samtools</span> flagstat results/bam/V1_S1_L001_ds_trim_align.bam</span></code></pre>
</div>
<p>This will give you the following statistics about your sorted bam
file:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>2280568 + 0 in total (QC-passed reads + QC-failed reads)
357208 + 0 secondary
0 + 0 supplementary
0 + 0 duplicates
1422795 + 0 mapped (62.39% : N/A)
1923360 + 0 paired in sequencing
961680 + 0 read1
961680 + 0 read2
1022006 + 0 properly paired (53.14% : N/A)
1039730 + 0 with itself and mate mapped
25857 + 0 singletons (1.34% : N/A)
15468 + 0 with mate mapped to a different chr
11429 + 0 with mate mapped to a different chr (mapQ&gt;=5)</code></pre>
</div>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>What do these different values mean? You might think to start with
<code>samtools flagstat --help</code>, but that just gives you the
options for the command. What could you do to find out more?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>There are two common approaches to this kind of challenge:</p>
<ol style="list-style-type: decimal"><li><p>It is a good idea to start with the full documentation or manual,
which is usually available online. The manual for samtools and
associated programs is <a href="https://www.htslib.org/doc/#manual-pages" class="external-link">from HTSlib here</a>,
but it is not immediately clear how to find the information on the
flagstat command. You can find <a href="https://www.htslib.org/doc/samtools-flagstat.html" class="external-link">the specific
manual page here</a> after a little poking around.</p></li>
<li><p>If the manual doesn’t help you, you can also try searching the
internet with your question. Good search terms are specific and clear.
In this case, you could try “samtools flagstat output”. The first hit
(at least for me) is the manual, but the second is <a href="https://www.biostars.org/p/84396/" class="external-link">a forum post</a> on Biostars
that might be helpful!</p></li>
</ol><p>If these two approaches don’t work, you can also talk to someone else
working on similar problems, make your own post on a help forum, or
reach out to the people who made the program (see the attribution or
citation)!</p>
</div>
</div>
</div>
</div>
<p>In order for us to do some of our next steps, we will need to index
the BAM file using <code>samtools</code>:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">$</span> samtools index results/bam/V1_S1_L001_ds_trim_align.bam</span></code></pre>
</div>
<p>This generates another file with the same name but a new extension,
.bai (bam index).</p>
</div>
</div>
</section><section><h2 class="section-heading" id="feature-read-counting">Feature (read) counting<a class="anchor" aria-label="anchor" href="#feature-read-counting"></a></h2>
<hr class="half-width"><p>Read counting is calculating the number of reads for a particular
sample that align to the reference at a given position, possibly at a
gene or part of a gene. Similar to other steps in this workflow, there
are a number of tools available for read counting. In this workshop we
will be using the <code>featureCounts</code> program within the <a href="https://subread.sourceforge.net/" class="external-link"><code>subread</code>
package</a>. <em>Note: a package is a word we use to describe a
collection of programs or scripts.</em></p>
<p>Our final output here will be a matrix with genes or other genomic
features as rows and samples as columns, with the number of reads
aligned to that feature in that sample in each cell.</p>
<figure><img src="fig/rna-seq-workflow.png" alt="workflow" class="figure mx-auto d-block"></figure><p>First, let’s find and load the module.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="ex">$</span> module avail subread</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>-------------------------- /share/apps/modulefiles/modules ---------------------------
   subread-2.0.2-gcc-10.2.0

Use "module spider" to find all possible modules and extensions.
Use "module keyword key1 key2 ..." to search for all possible modules matching any of
the "keys".</code></pre>
</div>
<p>Okay, let’s load the module and run it on our sorted, indexed bam
file. We need to give <code>featureCounts</code> an annotation file
after the <code>-a</code> flag. Annotation files store where different
genomic features like genes or parts of genes are start and end in the
genome. For this project, our annotation file is our <a href="https://en.wikipedia.org/wiki/Gene_transfer_format" class="external-link">gene transfer
format</a> (.gtf) genome file. We will also give it the output file name
after <code>-o</code>, tell it to use 24 threads (<code>-T 24</code>),
that the input is paired end (<code>-p</code>), and to only count read
pairs with both ends aligned (<code>-B</code>) to the same strand of the
same chromosome (<code>-C</code>). Remember that we can learn more about
these options with the manual or with <code>featureCounts --help</code>.
Finally, we tell <code>featureCounts</code> where to find the input file
<strong>or files</strong>.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="ex">$</span> module load subread-2.0.2-gcc-10.2.0</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="ex">$</span> featureCounts <span class="at">-a</span> data/genome/Homo_sapiens.GRCh38.111.gtf <span class="at">-o</span> results/counts/V1_S1_L001_counts.txt <span class="at">-T</span> 24 <span class="at">-p</span> <span class="at">-B</span> <span class="at">-C</span> results/bam/V1_S1_L001_ds_trim_align.bam</span></code></pre>
</div>
<p>You should see output that looks like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
        ==========     _____ _    _ ____  _____  ______          _____
        =====         / ____| |  | |  _ \|  __ \|  ____|   /\   |  __ \
          =====      | (___ | |  | | |_) | |__) | |__     /  \  | |  | |
            ====      \___ \| |  | |  _ &lt;|  _  /|  __|   / /\ \ | |  | |
              ====    ____) | |__| | |_) | | \ \| |____ / ____ \| |__| |
        ==========   |_____/ \____/|____/|_|  \_\______/_/    \_\_____/
	  v2.0.2

//========================== featureCounts setting ===========================\\
||                                                                            ||
||             Input files : 1 BAM file                                       ||
||                                                                            ||
||                           V1_S1_L001_ds_trim_align.bam                     ||
||                                                                            ||
||             Output file : V1_S1_L001_counts.txt                            ||
||                 Summary : V1_S1_L001_counts.txt.summary                    ||
||              Paired-end : yes                                              ||
||        Count read pairs : no                                               ||
||              Annotation : Homo_sapiens.GRCh38.111.gtf (GTF)                ||
||      Dir for temp files : results                                          ||
||                                                                            ||
||                 Threads : 24                                               ||
||                   Level : meta-feature level                               ||
||      Multimapping reads : not counted                                      ||
|| Multi-overlapping reads : not counted                                      ||
||   Min overlapping bases : 1                                                ||
||                                                                            ||
\\============================================================================//

//================================= Running ==================================\\
||                                                                            ||
|| Load annotation file Homo_sapiens.GRCh38.111.gtf ...                       ||
||    Features : 1650905                                                      ||
||    Meta-features : 63241                                                   ||
||    Chromosomes/contigs : 47                                                ||
||                                                                            ||
|| Process BAM file V1_S1_L001_ds_trim_align.bam...                           ||
||    Paired-end reads are included.                                          ||
||    The reads are assigned on the single-end mode.                          ||
||    Total alignments : 2280568                                              ||
||    Successfully assigned alignments : 709130 (31.1%)                       ||
||    Running time : 0.01 minutes                                             ||
||                                                                            ||
|| Write the final count table.                                               ||
|| Write the read assignment summary.                                         ||
||                                                                            ||
|| Summary of counting results can be found in file "results/V1_S1_L001_coun  ||
|| ts.txt.summary"                                                            ||
||                                                                            ||
\\============================================================================//</code></pre>
</div>
<p>And if you look in the <code>results/counts/</code> directory, you
can see two new files:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">$</span> ls results/counts</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>V1_S1_L001_counts.txt  V1_S1_L001_counts.txt.summary</code></pre>
</div>
<p>If we look at the summary file, we can learn a bit more about how
well this sample performed. You might want to look up what the different
values mean.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="ex">$</span> less results/counts/V1_S1_L001_counts.txt.summary</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Status  results/bam/V1_S1_L001_ds_trim_align.bam
Assigned        709130
Unassigned_Unmapped     857773
Unassigned_Read_Type    0
Unassigned_Singleton    0
Unassigned_MappingQuality       0
Unassigned_Chimera      0
Unassigned_FragmentLength       0
Unassigned_Duplicate    0
Unassigned_MultiMapping 509021
Unassigned_Secondary    0
Unassigned_NonSplit     0
Unassigned_NoFeatures   62313
Unassigned_Overlapping_Length   0
Unassigned_Ambiguity    142331</code></pre>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>Now that we have run through our workflow for a single sample, we
want to repeat this workflow for our other samples. However, we do not
want to type each of these individual steps again each time. That would
be very time consuming and error-prone, and would become impossible as
we gathered more and more samples. Luckily, we already know the tools we
need to use to automate this workflow and run it on as many files as we
want using a single line of code. Those tools are: wildcards, for loops,
bash scripts, and job submission with slurm. <strong>Your challenge is
to do this!</strong></p>
<p><em>Hint: Remember that <code>featureCounts</code> can take multiple
bam files as input, and will put them altogether in a single count
file.</em></p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>Your solution can take many forms, but you will know you have solved
it when you have a single count and count summary file in
<code>results/counts/</code> for all three samples, after running one or
more sbatch scripts!</p>
</div>
</div>
</div>
</div>
<p>Once you have all the samples aligned, compressed to bam, sorted,
indexed, and feature counted, we’re ready to do some data science!</p>
<div id="installing-software" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="installing-software" class="callout-inner">
<h3 class="callout-title">Installing software</h3>
<div class="callout-content">
<p>It is worth noting that all of the software we are using for this
workshop has been pre-installed on our remote server. Not every program
that you might need to use will be installed, however. You might need to
install the software, or to ask the people in charge of your remote
server to install it for you. It’s a good idea to find out how your
specific remote server managers prefer that you do this. There might be
a wiki or help page for the computer cluster, or you might need to email
someone. <a href="https://www.umb.edu/rc/hpc/" class="external-link">Chimera</a> is managed by
Jeff Dusenberry, the director of research computing at UMass Boston. You
can email him or another IT professional at <a href="mailto:It-rc@umb.edu" class="email">It-rc@umb.edu</a>. There is
also <a href="https://www.umb.edu/rc/applications/" class="external-link">a short
informational page</a> about applications (programs) on chimera.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>Bioinformatic command line tools are collections of commands that
can be used to carry out bioinformatic analyses.</li>
<li>To use most powerful bioinformatic tools, you will need to use the
command line.</li>
<li>There are many different file formats for storing genomics data. It
is important to understand what type of information is contained in each
file, and how it was derived.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="03-trimming.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="05-automation.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="03-trimming.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Trimming and
        </a>
        <a class="chapter-link float-end" href="05-automation.html" rel="next">
          Next: Analyzing Read Count...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/btmoyers/wrangling-transcriptomics/edit/main/episodes/04-variant_calling.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/btmoyers/wrangling-transcriptomics/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/btmoyers/wrangling-transcriptomics/" class="external-link">Source</a></p>
				<p><a href="https://github.com/btmoyers/wrangling-transcriptomics/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:brook.moyers@gmail.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.10" class="external-link">sandpaper (0.16.10)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.7" class="external-link">pegboard (0.7.7)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://btmoyers.github.io/wrangling-transcriptomics/04-variant_calling.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "RNA-Seq Workflow",
  "creativeWorkStatus": "active",
  "url": "https://btmoyers.github.io/wrangling-transcriptomics/04-variant_calling.html",
  "identifier": "https://btmoyers.github.io/wrangling-transcriptomics/04-variant_calling.html",
  "dateCreated": "2024-08-06",
  "dateModified": "2024-08-21",
  "datePublished": "2024-11-19"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

