<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Data Wrangling and Processing for Transcriptomics: Trimming and Filtering</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../03-trimming.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Data Wrangling and Processing for Transcriptomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Data Wrangling and Processing for Transcriptomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Data Wrangling and Processing for Transcriptomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 43%" class="percentage">
    43%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 43%" aria-valuenow="43" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../03-trimming.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-background.html">1. Introduction and Experimental Planning Considerations for RNA-Seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-quality-control.html">2. Assessing Read Quality</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        3. Trimming and Filtering
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#cleaning-reads">Cleaning reads</a></li>
<li><a href="#accessing-the-cutadapt-module">Accessing the cutadapt module</a></li>
<li><a href="#cutadapt-options">cutadapt options</a></li>
<li><a href="#running-cutadapt">Running cutadapt</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-variant_calling.html">4. Variant Calling Workflow</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-automation.html">5. Automating a Variant Calling Workflow</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/02-quality-control.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/04-variant_calling.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/02-quality-control.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Assessing Read
        </a>
        <a class="chapter-link float-end" href="../instructor/04-variant_calling.html" rel="next">
          Next: Variant Calling...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Trimming and Filtering</h1>
        <p>Last updated on 2024-08-19 |

        <a href="https://github.com/btmoyers/wrangling-transcriptomics/edit/main/episodes/03-trimming.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 55 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can I get rid of sequence data that does not meet my quality
standards?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Clean FASTQ reads using cutadapt.</li>
<li>Select and set multiple options for command-line bioinformatic
tools.</li>
<li>Write <code>for</code> loops with two variables.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="cleaning-reads">Cleaning reads<a class="anchor" aria-label="anchor" href="#cleaning-reads"></a>
<a class="anchor" aria-label="anchor" href="#cleaning-reads"></a></h2>
<hr class="half-width"><p>In the previous episode, we took a high-level look at the quality of
each of our samples using FastQC. We visualized per-base quality graphs
showing the distribution of read quality at each base across all reads
in a sample and extracted information about which samples have warnings
or failures for which quality checks.</p>
<p>It is very common to have some quality metrics fail or have some
moderately concerning values, and this may or may not be a problem for
your downstream application. For our RNA-Seq workflow, we can filter out
reads that have remnants from library preparation and sequencing or
remove some of the low quality sequences to reduce our false positive
rate due to sequencing error.</p>
<p>We will use a program called <a href="https://cutadapt.readthedocs.io/en/stable/index.html" class="external-link">cutadapt</a>
to filter poor quality reads and trim poor quality bases from our
samples.</p>
</section><section><h2 class="section-heading" id="accessing-the-cutadapt-module">Accessing the cutadapt module<a class="anchor" aria-label="anchor" href="#accessing-the-cutadapt-module"></a>
<a class="anchor" aria-label="anchor" href="#accessing-the-cutadapt-module"></a></h2>
<hr class="half-width"><p>Remember that we can look for modules (programs like this that might
be installed but that not everyone needs all the time) with
<code>module avail</code>:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> module avail cutadapt</span></code></pre>
</div>
<p>which on chimera should give you this output:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>-------------------------------- /share/apps/modulefiles/modules --------------------------------
   py-cutadapt-2.10-gcc-10.2.0-2x2ytr5

Use "module spider" to find all possible modules and extensions.
Use "module keyword key1 key2 ..." to search for all possible modules matching any of the
"keys".</code></pre>
</div>
<p>As it says, if you think something <strong>should</strong> be there
and it isn’t showing up, you can try <code>module spider</code>.</p>
<p>In our case, we will need to also load these modules to help cutadapt
work:</p>
<pre><code>module load py-dnaio-0.4.2-gcc-10.2.0-gaqzhv4
module load py-xopen-1.1.0-gcc-10.2.0-5kpnvqq
module load py-cutadapt-2.10-gcc-10.2.0-2x2ytr5</code></pre>
</section><section><h2 class="section-heading" id="cutadapt-options">cutadapt options<a class="anchor" aria-label="anchor" href="#cutadapt-options"></a>
<a class="anchor" aria-label="anchor" href="#cutadapt-options"></a></h2>
<hr class="half-width"><p>cutadapt has a variety of options to trim your reads. If we run the
following command, we can see some of our options.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">$</span> cutadapt <span class="at">--help</span></span></code></pre>
</div>
<p>Which will give you the following output:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>cutadapt version 2.10

Copyright (C) 2010-2020 Marcel Martin &lt;marcel.martin@scilifelab.se&gt;

cutadapt removes adapter sequences from high-throughput sequencing reads.

Usage:
    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq

For paired-end reads:
    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq

Replace "ADAPTER" with the actual sequence of your 3' adapter. IUPAC wildcard characters are supported. All reads from input.fastq will be written to output.fastq with the adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter sequences can be given (use further -a options), but only the best-matching adapter will be removed.

Input may also be in FASTA format. Compressed input and output is supported and auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for standard input/output. Without the -o option, output is sent to standard output.</code></pre>
</div>
<p>Plus a lot more options.</p>
<p>In paired end mode, cutadapt expects the two input files (R1 and R)
after the names of the output files given after the options
<code>-o</code> and <code>-p</code>. These files are described
below.</p>
<table class="table"><colgroup><col width="11%"><col width="88%"></colgroup><thead><tr class="header"><th>option</th>
<th>meaning</th>
</tr></thead><tbody><tr class="odd"><td>&lt;outputFile1&gt;</td>
<td>After -o, output file that contains trimmed or filtered reads from
the first input file (typically ‘R1’)</td>
</tr><tr class="even"><td>&lt;outputFile2&gt;</td>
<td>After -p, output file that contains trimmed or filtered reads from
the first input file (typically ‘R2’)</td>
</tr><tr class="odd"><td>&lt;inputFile1&gt;</td>
<td>Input reads to be trimmed. Typically the file name will contain an
<code>_1</code> or <code>_R1</code> in the name.</td>
</tr><tr class="even"><td>&lt;inputFile2&gt;</td>
<td>Input reads to be trimmed. Typically the file name will contain an
<code>_2</code> or <code>_R2</code> in the name.</td>
</tr></tbody></table><p>Cutadapt can remove adapter sequences that are in your sequence data,
trim or remove low-quality bases or reads, and do a few other useful
things. We will use only a few of these options and trimming steps in
our analysis. It is important to understand the steps you are using to
clean your data. For more information about the cutadapt arguments and
options, see <a href="https://cutadapt.readthedocs.io/en/stable/guide.html" class="external-link">the cutadapt
manual</a>.</p>
<p>However, a complete command for cutadapt will look something like the
command below. This command is an example and will not work, as we do
not have the files it refers to:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">$</span> cutadapt <span class="at">-a</span> AGATCGGAAGAGCACACGTCTGAACTCCAGTCA <span class="at">-A</span> AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT <span class="at">--trim-n</span> <span class="at">-m</span> 25 <span class="at">-o</span> example_R1_trim.fastq <span class="at">-p</span> example_R2_trim.fastq exampleR1.fastq exampleR2.fastq</span></code></pre>
</div>
<p>In this example, we have told cutadapt:</p>
<table class="table"><colgroup><col width="11%"><col width="88%"></colgroup><thead><tr class="header"><th>code</th>
<th>meaning</th>
</tr></thead><tbody><tr class="odd"><td><code>-a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA</code></td>
<td>identify and remove bases that match the Illumina adapter from each
R1 read</td>
</tr><tr class="even"><td><code>-A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT</code></td>
<td>identify and remove bases that match the Illumina adapter from each
R2 read</td>
</tr><tr class="odd"><td><code>--trim-n</code></td>
<td>trim any N bases from the beginning or end of each read</td>
</tr><tr class="even"><td><code>-m 25</code></td>
<td>discard any reads that are shorter than 25 bases after trimming</td>
</tr><tr class="odd"><td><code>-o example_R1_trim.fastq</code></td>
<td>the output file for trimmed and filtered reads from the R1 input
file</td>
</tr><tr class="even"><td><code>-p example_R2_trim.fastq</code></td>
<td>the output file for trimmed and filtered reads from the R2 input
file</td>
</tr><tr class="odd"><td><code>exampleR1.fastq</code></td>
<td>the input R1 fastq file</td>
</tr><tr class="even"><td><code>exampleR2.fastq</code></td>
<td>the input R2 fastq file</td>
</tr></tbody></table><div id="multi-line-commands" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="multi-line-commands" class="callout-inner">
<h3 class="callout-title">Multi-line commands<a class="anchor" aria-label="anchor" href="#multi-line-commands"></a>
</h3>
<div class="callout-content">
<p>Some of the commands we ran in this lesson are long! When typing a
long command into your terminal or nano, you can use the <code>\</code>
character to separate code chunks onto separate lines. This can make
your code more readable.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="running-cutadapt">Running cutadapt<a class="anchor" aria-label="anchor" href="#running-cutadapt"></a>
<a class="anchor" aria-label="anchor" href="#running-cutadapt"></a></h2>
<hr class="half-width"><p>Now we will run cutadapt on our data. To begin, navigate to your
<code>untrimmed_fastq</code> data directory:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">$</span> cd ~/1_project/data/untrimmed_fastq</span></code></pre>
</div>
<p>We are going to run cutadapt on one of our paired-end samples (V1).
We will identify and remove any leftover adapter sequences from the
reads. We will also remove N bases from the ends of the reads and filter
out any reads that are shorter than 25 bases in our trimmed sequences
(like in our example above).</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">$</span> cutadapt <span class="at">-a</span> AGATCGGAAGAGCACACGTCTGAACTCCAGTCA <span class="at">-A</span> AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT <span class="at">--trim-n</span> <span class="at">-m</span> 35 <span class="at">-o</span> V1_S1_L001_R1_001_ds_trim.fastq <span class="at">-p</span> V1_S1_L001_R2_001_ds_trim.fastq V1_S1_L001_R1_001_downsampled.fastq V1_S1_L001_R2_001_downsampled.fastq</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>This is cutadapt 2.10 with Python 3.8.6
Command line parameters: -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT --trim-n -m 35 -o V1_S1_L001_R1_001_ds_trim.fastq -p V1_S1_L001_R2_001_ds_trim.fastq V1_S1_L001_R1_001_downsampled.fastq V1_S1_L001_R2_001_downsampled.fastq
Processing reads on 1 core in paired-end mode ...
[         8=-] 00:00:20       962,229 reads  @     21.8 µs/read;   2.76 M reads/minute
Finished in 21.23 s (22 us/read; 2.72 M reads/minute).

=== Summary ===

Total read pairs processed:            962,229
  Read 1 with adapter:                  18,572 (1.9%)
  Read 2 with adapter:                  21,863 (2.3%)
Pairs written (passing filters):       961,614 (99.9%)

Total basepairs processed:    98,147,358 bp
  Read 1:    49,073,679 bp
  Read 2:    49,073,679 bp
Total written (filtered):     97,939,516 bp (99.8%)
  Read 1:    48,977,906 bp
  Read 2:    48,961,610 bp</code></pre>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise"></a>
</h3>
<div class="callout-content">
<p>Use the output from your cutadapt command to answer the following
questions.</p>
<ol style="list-style-type: decimal"><li>What percent of read pairs passed our filters?</li>
<li>What percent of basepairs passed our filters?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal"><li>99.9%</li>
<li>99.8%</li>
</ol></div>
</div>
</div>
</div>
<p>We can confirm that we have our output files:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">$</span> ls V1<span class="pp">*</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>V1_S1_L001_R1_001_downsampled.fastq  V1_S1_L001_R2_001_downsampled.fastq
V1_S1_L001_R1_001_ds_trim.fastq      V1_S1_L001_R2_001_ds_trim.fastq</code></pre>
</div>
<p>The output files are also FASTQ files. It might be smaller than our
input file, if we have removed reads. We can confirm this:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-thor</span> V1<span class="pp">*</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>-rw-rw-r-- 1 brook.moyers 146M Aug  7 00:55 V1_S1_L001_R1_001_downsampled.fastq
-rw-rw-r-- 1 brook.moyers 146M Aug  7 00:55 V1_S1_L001_R2_001_downsampled.fastq
-rw-rw-r-- 1 brook.moyers 146M Aug 16 17:22 V1_S1_L001_R1_001_ds_trim.fastq
-rw-rw-r-- 1 brook.moyers 146M Aug 16 17:22 V1_S1_L001_R2_001_ds_trim.fastq</code></pre>
</div>
<p>Hmmm, it doesn’t look that different (they are all 146 MB). Maybe
they are not that different because we kept most reads! We could check
the number of lines, maybe using a for loop.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">$</span> for filename in V1<span class="pp">*</span><span class="kw">;</span> <span class="cf">do</span> <span class="va">lines</span><span class="op">=</span><span class="va">$(</span><span class="fu">wc</span> <span class="at">-l</span> <span class="va">${filename})</span><span class="kw">;</span> <span class="bu">echo</span> <span class="va">$lines</span><span class="kw">;</span> <span class="cf">done</span></span></code></pre>
</div>
<pre><code>3848916 V1_S1_L001_R1_001_downsampled.fastq
3846456 V1_S1_L001_R1_001_ds_trim.fastq
3848916 V1_S1_L001_R2_001_downsampled.fastq
3846456 V1_S1_L001_R2_001_ds_trim.fastq</code></pre>
<p>Yes, it looks like the trimmed files are a couple thousand lines
shorter.</p>
<p>We have just successfully run cutadapt on one pair of our FASTQ
files! However, there is some bad news. cutadapt can only operate on one
sample at a time and we have more than one sample. The good news is that
we can use a script to iterate through our sample files quickly!</p>
<p>Here is the text of a script you can use to do it:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="co">#SBATCH --job-name=trim # you can give your job a name</span></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="co">#SBATCH --nodes 1 # the number of processors or tasks</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=2</span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a><span class="co">#SBATCH --account=itcga # our account</span></span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a><span class="co">#SBATCH --reservation=ITCGA_AUG2024 # this gives us special access during the workshop</span></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a><span class="co">#SBATCH --time=1:00:00 # the maximum time for the job</span></span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a><span class="co">#SBATCH --mem=4gb # the amount of RAM</span></span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a><span class="co">#SBATCH --partition=itcga # the specific server in chimera we are using</span></span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a><span class="co">#SBATCH --error=%x-%A.err   # a filename to save error messages into</span></span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a><span class="co">#SBATCH --output=%x-%A.out  # a filename to save any printed output into</span></span>
<span id="cb16-13"><a href="#cb16-13" tabindex="-1"></a></span>
<span id="cb16-14"><a href="#cb16-14" tabindex="-1"></a><span class="co"># Usage: sbatch cutadapt.sh path/to/input_dir path/to/output_dir</span></span>
<span id="cb16-15"><a href="#cb16-15" tabindex="-1"></a><span class="co"># Works for paired end reads where both end in the same *_001_downsampled.fastq</span></span>
<span id="cb16-16"><a href="#cb16-16" tabindex="-1"></a></span>
<span id="cb16-17"><a href="#cb16-17" tabindex="-1"></a><span class="co"># Module load</span></span>
<span id="cb16-18"><a href="#cb16-18" tabindex="-1"></a><span class="ex">module</span> load py-dnaio-0.4.2-gcc-10.2.0-gaqzhv4</span>
<span id="cb16-19"><a href="#cb16-19" tabindex="-1"></a><span class="ex">module</span> load py-xopen-1.1.0-gcc-10.2.0-5kpnvqq</span>
<span id="cb16-20"><a href="#cb16-20" tabindex="-1"></a><span class="ex">module</span> load py-cutadapt-2.10-gcc-10.2.0-2x2ytr5</span>
<span id="cb16-21"><a href="#cb16-21" tabindex="-1"></a></span>
<span id="cb16-22"><a href="#cb16-22" tabindex="-1"></a><span class="co"># Define variables</span></span>
<span id="cb16-23"><a href="#cb16-23" tabindex="-1"></a><span class="va">input_dir</span><span class="op">=</span><span class="va">$1</span> <span class="co"># takes this from the command line, first item after the script</span></span>
<span id="cb16-24"><a href="#cb16-24" tabindex="-1"></a><span class="va">output_dir</span><span class="op">=</span><span class="va">$2</span> <span class="co"># takes this from the command line, second item</span></span>
<span id="cb16-25"><a href="#cb16-25" tabindex="-1"></a></span>
<span id="cb16-26"><a href="#cb16-26" tabindex="-1"></a><span class="cf">for</span> R1_fastq <span class="kw">in</span> <span class="va">${input_dir}</span>/<span class="pp">*</span>_R1_001_downsampled.fastq</span>
<span id="cb16-27"><a href="#cb16-27" tabindex="-1"></a> <span class="cf">do</span></span>
<span id="cb16-28"><a href="#cb16-28" tabindex="-1"></a> <span class="co"># Pull basename</span></span>
<span id="cb16-29"><a href="#cb16-29" tabindex="-1"></a> <span class="va">name</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">${R1_fastq}</span> _R1_001_downsampled.fastq<span class="va">)</span></span>
<span id="cb16-30"><a href="#cb16-30" tabindex="-1"></a></span>
<span id="cb16-31"><a href="#cb16-31" tabindex="-1"></a> <span class="co"># Run cutadapt</span></span>
<span id="cb16-32"><a href="#cb16-32" tabindex="-1"></a> <span class="ex">cutadapt</span> <span class="at">-a</span> AGATCGGAAGAGCACACGTCTGAACTCCAGTCA <span class="dt">\</span></span>
<span id="cb16-33"><a href="#cb16-33" tabindex="-1"></a> <span class="at">-A</span> AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT <span class="at">--trim-n</span> <span class="at">-m</span> 25 <span class="dt">\</span></span>
<span id="cb16-34"><a href="#cb16-34" tabindex="-1"></a> <span class="at">-o</span> <span class="va">${output_dir}</span>/<span class="va">${name}</span>_R1_001_ds_trim.fastq <span class="dt">\</span></span>
<span id="cb16-35"><a href="#cb16-35" tabindex="-1"></a> <span class="at">-p</span> <span class="va">${output_dir}</span>/<span class="va">${name}</span>_R2_001_ds_trim.fastq <span class="dt">\</span></span>
<span id="cb16-36"><a href="#cb16-36" tabindex="-1"></a> <span class="va">${input_dir}</span>/<span class="va">${R1_fastq}</span> <span class="dt">\</span></span>
<span id="cb16-37"><a href="#cb16-37" tabindex="-1"></a> <span class="va">${input_dir}</span>/<span class="va">${name}</span>_R2_001_downsampled.fastq</span>
<span id="cb16-38"><a href="#cb16-38" tabindex="-1"></a></span>
<span id="cb16-39"><a href="#cb16-39" tabindex="-1"></a> <span class="bu">echo</span> cutadapt is finished with <span class="va">$name</span></span>
<span id="cb16-40"><a href="#cb16-40" tabindex="-1"></a></span>
<span id="cb16-41"><a href="#cb16-41" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise-1"></a>
</h3>
<div class="callout-content">
<p>Take a second to look, really look, through the script and note what
each line does. There are a few new elements here that you might not
have encountered before: what could you do to understand them?</p>
<ol style="list-style-type: decimal"><li>Write down in plain English what this script does, line by
line.</li>
<li>How could you modify this script if your FASTQ files ended in
<code>_1.fastq</code> and <code>_2.fastq</code> instea?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<ol style="list-style-type: decimal"><li><p>Your answer may vary, but here is one description: This is a bash
shell script that can run for up to an hour on the itcga partition and
reservation with 1 node, 2 CPUs, and 4 GB of memory. First, it loads
three modules necessary to run cutadapt, then stores the paths that come
after the name of the script on the command line as input_dir and
output_dir. Then for each file in the input_dir that ends with
<code>_R1_001_downsampled.fastq</code>, it runs cutadapt with it and its
paired R2 to remove adapters, trim Ns from the ends of reads, and filter
out any reads shorter than 25 bases after trimming. It stores the output
with the file extension <code>001_ds_trim.fastq</code> in the
output_dir.</p></li>
<li><p>You will need to modify the for loop call, the basename call, and
the cutadapt call as follows:</p></li>
</ol><div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="cf">for</span> R1_fastq <span class="kw">in</span> <span class="va">${input_dir}</span>/<span class="pp">*</span>_1.fastq</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="va">name</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">${R1_fastq}</span> _1.fastq<span class="va">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a> <span class="ex">cutadapt</span> <span class="at">-a</span> AGATCGGAAGAGCACACGTCTGAACTCCAGTCA <span class="dt">\</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a> <span class="at">-A</span> AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT <span class="at">--trim-n</span> <span class="at">-m</span> 25 <span class="dt">\</span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a> <span class="at">-o</span> <span class="va">${output_dir}</span>/<span class="va">${name}</span>_R1_001_ds_trim.fastq <span class="dt">\</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a> <span class="at">-p</span> <span class="va">${output_dir}</span>/<span class="va">${name}</span>_R2_001_ds_trim.fastq <span class="dt">\</span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a> <span class="va">${input_dir}</span>/<span class="va">${R1_fastq}</span> <span class="dt">\</span></span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a> <span class="va">${input_dir}</span>/<span class="va">${name}</span>_2.fastq</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>It is common that you might start your work with a shared script or
code from a collaborator or someone else. This kind of exercise can save
you a lot of headaches when trying to adapt it to your own workflow.</p>
<p>Okay, now that you really understand this script, let’s run it:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">$</span> mkdir ~/1_project/data/trimmed_fastq</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a><span class="ex">$</span> sbatch cutadapt.sh ~/1_project/data/untrimmed_fastq ~/1_project/data/trimmed_fastq</span></code></pre>
</div>
<p>If you check the trimmed_fastq directory, you can see that your
trimmed fastq files have been stored there!</p>
<p>You might also notice that your log files are kind of annoyingly
piling up in various places. Think about possible ways to manage where
they are stored!</p>
<p>We have now completed the trimming and filtering steps of our quality
control process! Before we move on, let’s move our trimmed FASTQ files
to a new subdirectory within our <code>data/</code> directory.</p>
<div id="bonus-exercise-advanced" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="bonus-exercise-advanced" class="callout-inner">
<h3 class="callout-title">Bonus exercise (advanced)<a class="anchor" aria-label="anchor" href="#bonus-exercise-advanced"></a>
</h3>
<div class="callout-content">
<p>Now that our samples have gone through quality control, they should
perform better on the quality tests run by FastQC. Go ahead and re-run
FastQC on your trimmed FASTQ files and visualize the HTML files to see
whether your per base sequence quality is higher after trimming. Hint:
you might need to modify your <code>fastqc.sh</code> script.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>There are a few ways to do this, but this one will work! Modify your
<code>fastqc.sh</code> script like this:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="co"># module load</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="ex">module</span> load fastqc-0.11.9-gcc-10.2.0-osi6pqc</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a><span class="co"># Define variables</span></span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a><span class="va">input_dir</span><span class="op">=</span><span class="va">$1</span> </span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a><span class="co"># run fastqc</span></span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a><span class="ex">fastqc</span> <span class="va">${input_dir}</span>/<span class="pp">*</span>.fastq</span></code></pre>
</div>
<p>Then you can run it by specifying the input directory on the command
line.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="ex">$</span> sbatch fastqc.sh ~/1_project/data/trimmed_fastq/</span></code></pre>
</div>
<p>After trimming and filtering, our overall quality is much higher, we
have a distribution of sequence lengths, and more samples pass adapter
content. However, you may want to explore the other options that <a href="https://cutadapt.readthedocs.io/en/stable/" class="external-link">cutadapt</a> includes
to refine your quality filtering.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>The options you set for the command-line tools you use are
important!</li>
<li>Data cleaning is an essential step in a genomics workflow.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/02-quality-control.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/04-variant_calling.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/02-quality-control.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Assessing Read
        </a>
        <a class="chapter-link float-end" href="../instructor/04-variant_calling.html" rel="next">
          Next: Variant Calling...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/btmoyers/wrangling-transcriptomics/edit/main/episodes/03-trimming.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/btmoyers/wrangling-transcriptomics/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/btmoyers/wrangling-transcriptomics/" class="external-link">Source</a></p>
				<p><a href="https://github.com/btmoyers/wrangling-transcriptomics/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:brook.moyers@gmail.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.5" class="external-link">sandpaper (0.16.5)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.6" class="external-link">pegboard (0.7.6)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.3" class="external-link">varnish (1.0.3)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://btmoyers.github.io/wrangling-transcriptomics/instructor/03-trimming.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Trimming and Filtering",
  "creativeWorkStatus": "active",
  "url": "https://btmoyers.github.io/wrangling-transcriptomics/instructor/03-trimming.html",
  "identifier": "https://btmoyers.github.io/wrangling-transcriptomics/instructor/03-trimming.html",
  "dateCreated": "2024-08-06",
  "dateModified": "2024-08-19",
  "datePublished": "2024-08-19"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

